controllers/matchControllers.js

```diff
@@ -9,22 +13,154 @@ exports.getMatch = async (req, res, next) => {
+    return;
+  }
 
-  //todo : add validation to see if these two books havent already been matched
+  // check if user owns one book and is interested in one book
+  const userIsInterestedInBook =
+    user.booksInterestedIn.includes(bookOne) ||
+    user.booksInterestedIn.includes(bookTwo);
+
+  const userOffersBook =
+    user.booksToOffer.includes(bookOne) || user.booksToOffer.includes(bookTwo);
+
+  if (!userIsInterestedInBook || !userOffersBook) {
+    next(customError(`Match is not valid`, 400));
+    return;
+  }
+
+  const checkIfMatchExists = () => {
+    for (userMatch in user.matches) {
+      if (
+        userMatch.bookOne === (bookOne || bookTwo) &&
+        userMatch.bookTwo === (bookOne || bookTwo)
+      )
+        console.log('books already matched');
+      return true;
+    }
+    return false;
+  };
 
   try {
-    let newMatch = await Match.create(matchData);
-    res.json(newMatch);
+    const createMatch = async (data) => {
+      let newMatch = await Match.create(req.body)
+        .populate('bookOne')
+        .populate('bookTwo');
+
+      await User.findByIdAndUpdate(bookOne.owner, {
+        $push: { matches: newMatch._id },
+      });
+      await User.findByIdAndUpdate(bookTwo.owner, {
+        $push: { matches: newMatch._id },
+      });
+
+      res.json(newMatch);
+    };
+
+    if (user.matches.length < 1) {
+      createMatch(req.body);
+    } else {
+      // check if matches contain these books already = true/false
+
+      console.log('does match exist?', checkIfMatchExists());
+
+      if (checkIfMatchExists()) {
+        next(customError(`Books already matched`, 400));
+        return;
+      } else {
+        createMatch(req.body);
+      }
+    }
+  } catch (err) {
+    if (err instanceof mongoose.Error.CastError) {
+      next(customError(`ID: ${id} is not valid`, 400));
+      next(err);
+    }
+  }
+};
+
+exports.updateMatch = async (req, res, next) => {
+  const { id } = req.params;
+
+  try {
+    const match = await Match.findById(id);
+
+    if (!match) {
+      next(customError(`Match with ID: ${id} does not exist`, 400));
+    }
+
+    Object.assign(match, req.body);
+    const updatedMatch = await match.save();
+    console.log('match updated');
+    res.json(updatedMatch);
+  } catch (err) {
+    if (err instanceof mongoose.Error.CastError) {
+      next(customError(`ID: ${id} is not valid`, 400));
+    }
+    next(err);
+  }
+};
+
+// delete Match
+exports.deleteMatch = async (req, res, next) => {
+  const { id } = req.params;
+
+  try {
+    const match = await Match.findById(id)
+      .populate('bookOne')
+      .populate('bookTwo');
+
+    if (!match) {
+      next(customError(`Match with ID: ${id} does not exist`, 400));
+      return;
+    }
+
+    await User.findByIdAndUpdate(match.bookOne.owner, {
+      $pull: { matches: match._id },
+    });
+    await User.findByIdAndUpdate(match.bookTwo.owner, {
+      $pull: { matches: match._id },
+    });
+    match.delete();
```

Comment:
await match.delete() ?

Reply:
true, works without as well, but you are completely right